engine/storage/snapshot/src/main/java/org/apache/cloudstack/storage/vmsnapshot/KvmFileBasedStorageVmSnapshotStrategy.java

```diff
@@ -0,0 +1,689 @@
+            snapshotDao.update(snapshotVO.getId(), snapshotVO);
+        }
+
+        vmSnapshotDetailsDao.removeDetails(oldParent.getId());
+
+        oldParent.setRemoved(DateUtil.now());
+        vmSnapshotDao.update(oldParent.getId(), oldParent);
+
+        transitStateWithoutThrow(oldParent, VMSnapshot.Event.ExpungeRequested);
+        transitStateWithoutThrow(oldParent, VMSnapshot.Event.OperationSucceeded);
+    }
+
+    @Override
+    public StrategyPriority canHandle(VMSnapshot vmSnapshot) {
+        if (!VMSnapshot.State.Allocated.equals(vmSnapshot.getState())) {
+            List<VMSnapshotDetailsVO> vmSnapshotDetails = vmSnapshotDetailsDao.findDetails(vmSnapshot.getId(), KVM_FILE_BASED_STORAGE_SNAPSHOT);
+            if (CollectionUtils.isEmpty(vmSnapshotDetails)) {
+                logger.debug("KVM file based storage VM snapshot strategy cannot handle [{}] as it is not a KVM file based storage VM snapshot.",
+                        vmSnapshot.getUuid());
+                return StrategyPriority.CANT_HANDLE;
+            }
+            return StrategyPriority.HIGHEST;
+        }
+
+        long vmId = vmSnapshot.getVmId();
+        boolean memorySnapshot = VMSnapshot.Type.DiskAndMemory.equals(vmSnapshot.getType());
+        return canHandle(vmId, null, memorySnapshot);
+    }
+
+    @Override
+    public StrategyPriority canHandle(Long vmId, Long rootPoolId, boolean snapshotMemory) {
+        VirtualMachine vm = userVmDao.findById(vmId);
+
+        String cantHandleLog = String.format("KVM file based storage VM snapshot strategy cannot handle VM snapshot for [%s]", vm);
+        if (snapshotMemory) {
+            logger.debug("{} as a snapshot with memory was requested.", cantHandleLog);
+            return StrategyPriority.CANT_HANDLE;
+        }
+
+        if (!Hypervisor.HypervisorType.KVM.equals(vm.getHypervisorType())) {
+            logger.debug("{} as the hypervisor is not KVM.", cantHandleLog);
+            return StrategyPriority.CANT_HANDLE;
+        }
+
+        if (CollectionUtils.isNotEmpty(vmSnapshotDao.findByVmAndByType(vmId, VMSnapshot.Type.DiskAndMemory))) {
+            logger.debug("{} as there is already a VM snapshot with disk and memory.", cantHandleLog);
+            return StrategyPriority.CANT_HANDLE;
+        }
+
+        List<VolumeVO> volumes = volumeDao.findByInstance(vmId);
+        for (VolumeVO volume : volumes) {
+            StoragePoolVO storagePoolVO = storagePool.findById(volume.getPoolId());
+            if (!supportedStoragePoolTypes.contains(storagePoolVO.getPoolType())) {
+                logger.debug(String.format("%s as the VM has a volume that is in a storage with unsupported type [%s].", cantHandleLog, storagePoolVO.getPoolType()));
+                return StrategyPriority.CANT_HANDLE;
+            }
+            List<SnapshotVO> snapshots = snapshotDao.listByVolumeIdAndTypeNotInAndStateNotRemoved(volume.getId(), Snapshot.Type.GROUP);
```

Comment:
@JoaoJandre, when I create a few VM snapshots with no memory, then try to take a volume snapshot, the request fails as expected, but a snapshot record is created with an Allocated state. If the user forgets to delete this snapshot record, there is no option to create more VM snapshots, and the error is unclear - `Failed to create vm snapshot: No strategy was able to handle requested snapshot for VM [5d96d7df-9b94-407a-8aab-a022dcae8cfe].`

Reply:
Thanks, @JoaoJandre, that will be the better option